core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java

```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
             }
         }
     }
+
+    private enum UnitDecimal
+    {
+        BYTE(BigDecimal.ONE),
+        KILOBYTE(new BigDecimal(1_000L)),
+        MEGABYTE(new BigDecimal(1_000_000L)),
+        GIGABYTE(new BigDecimal(1_000_000_000L)),
+        TERABYTE(new BigDecimal(1_000_000_000_000L)),
+        PETABYTE(new BigDecimal(1_000_000_000_000_000L)),
+        EXABYTE(new BigDecimal(1_000_000_000_000_000_000L)),
+        ZETTABYTE(new BigDecimal(1_000_000_000_000_000_000L).multiply(new BigDecimal(1_000L))),
+        YOTTABYTE(new BigDecimal(1_000_000_000_000_000_000L).multiply(new BigDecimal(1_000_000L)));
+
+        private final BigDecimal factor;
+
+        UnitDecimal(BigDecimal factor)
+        {
+            this.factor = factor;
+        }
+
+        public BigDecimal getFactor()
+        {
+            return factor;
+        }
+
+        public static UnitDecimal parse(String unitString, String dataSize)
+        {
+            switch (unitString) {
+                case "B":
+                    return BYTE;
+                case "kB":
+                    return KILOBYTE;
+                case "MB":
+                    return MEGABYTE;
+                case "GB":
+                    return GIGABYTE;
+                case "TB":
+                    return TERABYTE;
+                case "PB":
+                    return PETABYTE;
+                case "EB":
+                    return EXABYTE;
+                case "ZB":
+                    return ZETTABYTE;
+                case "YB":
+                    return YOTTABYTE;
+                default:
+                    throw invalidDataSize(dataSize);
+            }
+        }
+    }
+
+    @Description("Formats data size as bytes using a unit symbol")
+    @ScalarFunction(value = "format_data_size")
+    @SqlType(StandardTypes.VARCHAR)
+    public Slice formatDataSize(@SqlType(StandardTypes.BIGINT) long value)
+    {
+        return utf8Slice(formatLong((double) value, 1024));
+    }
+
+    @Description("Formats data size as bytes using a unit symbol")
+    @ScalarFunction(value = "format_data_size")
+    @SqlType(StandardTypes.VARCHAR)
+    public Slice formatDataSize(@SqlType(StandardTypes.DOUBLE) double value)
+    {
+        return utf8Slice(formatLong(value, 1024));
+    }
+
+    @Description("Formats data size as bytes using a unit symbol and a decimal base")
+    @ScalarFunction(value = "format_data_size_decimal")
+    @SqlType(StandardTypes.VARCHAR)
+    public Slice formatDataSizeDecimal(@SqlType(StandardTypes.BIGINT) long value)
+    {
+        return utf8Slice(formatLong((double) value, 1000));
+    }
+
+    @Description("Formats data size as bytes using a unit symbol and a decimal base")
+    @ScalarFunction(value = "format_data_size_decimal")
+    @SqlType(StandardTypes.VARCHAR)
+    public Slice formatDataSizeDecimal(@SqlType(StandardTypes.DOUBLE) double value)
```

Comment:
What does this method return when invoked with +/- infinity?
What does it return when invoked with NaN?

Reply:
Thanks, I missed this. I added conditions to avoid formatting these values (avoid adding a suffix, since `InfinityB` or `NaNB` doesn't make sense) and added test cases.